watchtower-plugin/src/retrier.rs

```diff
@@ -16,174 +16,248 @@ use crate::wt_client::WTClient;
-
-            let retrier = Retrier::new(self.wt_client.clone(), locator);
-            e.insert(retrier.clone());
-
-            Some(retrier)
+            e.insert(Retrier::new(self.wt_client.clone(), tower_id, locator));
         } else {
             log::debug!(
                 "Adding pending appointment {} to existing tower {}",
                 locator,
                 tower_id
             );
-            retriers
+            self.retriers
                 .get(&tower_id)
                 .unwrap()
                 .pending_appointments
                 .lock()
                 .unwrap()
                 .insert(locator);
-
-            None
         }
     }
+
+    /// Starts the given retrier.
+    fn start_retrying(&self, retrier: &Retrier) {
+        log::info!("Retrying tower {}", retrier.tower_id);
+        Retrier::retry(
+            retrier.clone(),
+            self.max_elapsed_time_secs,
+            self.max_interval_time_secs,
+        );
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum RetrierStatus {
+    /// Retrier is currently retrying the tower. If the retrier receives new appointments, it will
+    /// **try** to send them too (but it might not send it).
+    ///
+    /// If a retrier status is `Running`, then its associated tower is temporary unreachable.
+    Running,
+    /// Retrier finished retrying the tower successfully. Note that the retrier might receive new
+    /// appointments after it has finished successfully. In such a case, it should be re-started.
+    ///
+    /// If a retrier status is `Success`, then its associated tower is reachable.
+    Success,
+    /// Retrier failed retrying the tower.
+    ///
+    /// If a retrier status is `Failure`, then its associated tower is neither reachable nor temporary unreachable.
+    Failure,
 }
 
 #[derive(Clone)]
 pub struct Retrier {
     wt_client: Arc<Mutex<WTClient>>,
+    tower_id: TowerId,
     pending_appointments: Arc<Mutex<HashSet<Locator>>>,
+    status: Arc<Mutex<RetrierStatus>>,
 }
 
 impl Retrier {
-    pub fn new(wt_client: Arc<Mutex<WTClient>>, locator: Locator) -> Self {
+    pub fn new(wt_client: Arc<Mutex<WTClient>>, tower_id: TowerId, locator: Locator) -> Self {
         Self {
             wt_client,
+            tower_id,
             pending_appointments: Arc::new(Mutex::new(HashSet::from([locator]))),
+            status: Arc::new(Mutex::new(RetrierStatus::Success)),
         }
     }
 
-    async fn retry_tower(&self, tower_id: TowerId) -> Result<(), Error<&'static str>> {
+    fn has_pending_appointments(&self) -> bool {
+        !self.pending_appointments.lock().unwrap().is_empty()
+    }
+
+    fn set_status(&self, status: RetrierStatus) {
+        *self.status.lock().unwrap() = status;
+    }
+
+    pub fn is_running(&self) -> bool {
+        *self.status.lock().unwrap() == RetrierStatus::Running
+    }
+
+    pub fn ready(&self) -> bool {
```

Comment:
I kept it like so to make it inclusive for the two cases:
1- when a fresh new retrier is ready to start.
2- when a finished retrier is ready to restart.
needs_restart accounts for case 2 only. agree there might be more descriptive naming, but I want it to kinda include both cases.

Reply:
Fair enough, but then, at the very least, the docs need to be updated. Right now they only read about re-starting.

Notice this only works because you are overloading `RetrierStatus::Success` for just created and completed retriers, if you end up adding a new variant to the enum it'll need to be updated